### PR TITLE
Import Licence version purpose conditions

### PR DESCRIPTION
**Pull request by jonathangoulding**
https://eaflood.atlassian.net/browse/WATER-4649

We need to replace the import service logic to import a licence from NALD.

We will leave the import service running to add the NALD data in the import tables, but we will slowly replace the import-licence functionality and move this into the system repo.

This change will add the licence versions purpose conditions to the import logic.

Previous work that added licence version and licence version purpose - https://github.com/DEFRA/water-abstraction-system/pull/1195

### CODE REVIEW

File: app/services/import/legacy/transform-licence-version-purpose-conditions.service.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+async function go (regionCode, naldLicenceId, transformedLicence) {
+  const naldLicenceVersionPurposesConditions = await
+  FetchLicenceVersionPurposeConditionsService.go(regionCode, naldLicenceId)
+
+  for (const licenceVersion of transformedLicence.licenceVersions) {
+    for (const licenceVersionPurpose of licenceVersion.licenceVersionPurposes) {
+      const matchingLicenceVersionPurposeConditions =
+        _conditionsForLicenceVersionPurpose(licenceVersionPurpose, naldLicenceVersionPurposesConditions)
+
+      //  validate here -
```
Comment (jonathangoulding) [inline app/services/import/legacy/transform-licence-version-purpose-conditions.service.js:31]: Add validation

File: app/services/import/legacy/process-licence.service.js [hunk 2]
```diff
@@ -29,6 +30,7 @@ async function go (licenceRef) {
     // Pass the transformed licence through each transformation step, building the licence as we go
     await TransformLicenceVersionsService.go(regionCode, naldLicenceId, transformedLicence)
     await TransformLicenceVersionPurposesService.go(regionCode, naldLicenceId, transformedLicence)
+    await TransformLicenceVersionPurposeConditionsService.go(regionCode, naldLicenceId, transformedLicence)
```
Comment (jonathangoulding) [inline app/services/import/legacy/process-licence.service.js:33]: Add persistence

Comment (jonathangoulding) [reply app/services/import/legacy/process-licence.service.js:33]: Update tests

File: app/services/import/persist-licence.service.js [hunk 3]
```diff
@@ -97,10 +98,40 @@ async function _persistLicenceVersionPurpose (trx, updatedAt, licenceVersionPurp
...
+    const { id } = await _persistLicenceVersionPurpose(
+      trx, updatedAt, licenceVersionPurpose, licenceVersionId)
+
+    await _persistLicenceVersionPurposesConditions(trx, updatedAt, licenceVersionPurposeConditions, id)
   }
 }
 
+async function _persistLicenceVersionPurposesConditions (
+  trx, updatedAt, licenceVersionPurposeConditions, licenceVersionPurposeId) {
+  for (const licenceVersionPurposeCondition of licenceVersionPurposeConditions) {
```
Comment (jonathangoulding) [inline app/services/import/persist-licence.service.js:112]: This potentially can be done in a batch but i have followed the previous code pattern.

File: app/services/import/persist-licence.service.js [hunk 4]
```diff
@@ -97,10 +98,40 @@ async function _persistLicenceVersionPurpose (trx, updatedAt, licenceVersionPurp
...
+    const licenceVersionPurposeConditions = licenceVersionPurpose.licenceVersionPurposeConditions
+
+    const { id } = await _persistLicenceVersionPurpose(
+      trx, updatedAt, licenceVersionPurpose, licenceVersionId)
+
+    await _persistLicenceVersionPurposesConditions(trx, updatedAt, licenceVersionPurposeConditions, id)
   }
 }
 
+async function _persistLicenceVersionPurposesConditions (
```
Comment (Cruikshanks) [inline app/services/import/persist-licence.service.js:110]:
```suggestion
async function _persistLicenceVersionPurposeConditions (
```

File: app/services/import/persist-licence.service.js [hunk 5]
```diff
@@ -97,10 +98,40 @@ async function _persistLicenceVersionPurpose (trx, updatedAt, licenceVersionPurp
...
+  trx, updatedAt, licenceVersionPurposeConditions, licenceVersionPurposeId) {
+  for (const licenceVersionPurposeCondition of licenceVersionPurposeConditions) {
+    await _persistLicenceVersionPurposeConditions(
+      trx, updatedAt, licenceVersionPurposeCondition, licenceVersionPurposeId)
+  }
+}
+
+async function _persistLicenceVersionPurposeConditions (
```
Comment (Cruikshanks) [inline app/services/import/persist-licence.service.js:118]:
```suggestion
async function _persistLicenceVersionPurposeCondition (
```

File: app/validators/import/licence-version-purpose-condition.validator.js [hunk 6]
```diff
@@ -16,25 +16,21 @@ const Joi = require('joi')
...
-  })
+  const schema =
+    Joi.object({
+      licenceVersionPurposeConditionTypeId: Joi.string().guid().required(),
+      param1: Joi.string().allow(null).optional(),
+      param2: Joi.string().allow(null).optional(),
+      notes: Joi.string().allow(null).optional(),
+      externalId: Joi.string().required()
+    }
+    )
```
Comment (Cruikshanks) [inline app/validators/import/licence-version-purpose-condition.validator.js:27]:
```suggestion
    })
```

**Review by Cruikshanks — CHANGES_REQUESTED** [review]
First pass of corrections and suggested changes.

**Review by Cruikshanks — DISMISSED** [review]

**Review by Cruikshanks — APPROVED** [review]

**Pull request merged**